src/Data/Store/Impl.hs

```diff
@@ -165,27 +165,42 @@ decodeIO = decodeIOWith peek
 -- exceptions, and taking a 'Peek' to run. It is an exception to not
 -- consume all input.
 decodeIOWith :: Peek a -> BS.ByteString -> IO a
-decodeIOWith mypeek bs = do
-    (offset, x) <- decodeIOPortionWith mypeek bs
-    let remaining = BS.length bs - offset
-    if remaining > 0
-        then throwIO $ PeekException remaining "Didn't consume all input."
-        else return x
+decodeIOWith mypeek (BS.PS x s len) =
+    withForeignPtr x $ \ptr0 ->
+        let ptr = ptr0 `plusPtr` s
+        in decodeIOWithFromPtr mypeek ptr len
 {-# INLINE decodeIOWith #-}
 
 -- | Similar to 'decodeExPortionWith', but runs in the 'IO' monad.
 decodeIOPortionWith :: Peek a -> BS.ByteString -> IO (Offset, a)
 decodeIOPortionWith mypeek (BS.PS x s len) =
     withForeignPtr x $ \ptr0 ->
         let ptr = ptr0 `plusPtr` s
-            end = ptr `plusPtr` len
-         in do
-             (ptr2, x') <- runPeek mypeek end ptr
-             if ptr2 > end
-                 then throwIO $ PeekException (end `minusPtr` ptr2) "Overshot end of buffer"
-                 else return (ptr2 `minusPtr` ptr, x')
+        in decodeIOPortionWithFromPtr mypeek ptr len
 {-# INLINE decodeIOPortionWith #-}
 
+-- | Like 'decodeIOWith', but using 'Ptr' and length instead of a 'ByteString'.
+decodeIOWithFromPtr :: Peek a -> Ptr Word8 -> Int -> IO a
+decodeIOWithFromPtr mypeek ptr len = do
+    (offset, x) <- decodeIOPortionWithFromPtr mypeek ptr len
+    let remaining = len - offset
+    if remaining > 0
```

Comment:
I realize this is how the original code was written, but it seems slightly strange. Assuming I understand the goal here, I would have gone with:

```haskell
if len /= offset
```

This is better since:

* It also catches a case of consuming too much input (though that's probably better handled via an `assert` if we actually care about it)
* It's marginally faster (one comparison instead of a subtraction + comparison)
* Subjectively, I think it's clearer